### DISCUSSION TITLE
Project-Wide Customizable Default Settings for New Collection Field Names

### DISCUSSION THREAD

**Discussion by mkungla** [directus/directus#18824] [Feature Requests]
### Summary

While working with Directus, I noticed that when creating a new collection, the field names are hardcoded in the new-collection.vue file (source: [GitHub](https://github.com/directus/directus/blob/5a0032f606f40fadce496c8085b29a140418d277/app/src/modules/settings/routes/data-model/new-collection.vue#L143)). These include timestamp fields such as 'created_on', 'user_created', 'date_updated', 'user_updated' among others.

I believe that a customizable, project-wide default setting for these field names would greatly enhance user experience. With this feature, users could establish their preferred naming conventions for these fields under project settings, thus eliminating the need to manually rename these fields for each new collection.

This feature request was initially discussed in the #internal-apps channel and I'm submitting it here for further consideration.

### Motivation

The primary motivation for this feature request is to enhance the flexibility and efficiency of the Directus platform. Currently, users need to manually rename the default fields for each new collection they create, which can be time-consuming and prone to inconsistencies.

By allowing users to establish project-wide default field names, we can cater to diverse naming conventions and preferences. This feature would streamline the creation of new collections, enabling users to maintain a consistent schema across their project without the need for repetitive manual adjustments.

Moreover, this feature could extend beyond the existing default fields like status, sort, created_on, created_by, updated_on, updated_by, potentially offering the option to add other project-wide default fields. These could then be easily included when creating a new collection, further enhancing the user experience and the platform's adaptability.

This solution aims to eliminate the constraints imposed by hardcoded field names, fostering a more user-centric and customizable environment in Directus.

### Detailed Design

The proposed change focuses on modifying the behavior of the new-collection.vue file (source: [GitHub](https://github.com/directus/directus/blob/5a0032f606f40fadce496c8085b29a140418d277/app/src/modules/settings/routes/data-model/new-collection.vue#L143)).

In the current design, the defaultSystemFields are hardcoded, which limits customization. I propose that these values should instead be dynamically fetched from the project settings through the admin API, rather than being statically defined.

Here's a high-level illustration of the design:

- The user sets their preferred default field names in the project settings.
- When a new collection is being created, Directus calls the admin API to fetch these settings.
- The new-collection.vue file uses these settings to pre-fill the default fields.
- If desired, the user can still manually override these pre-filled values on a per-collection basis like currently we do it.

This approach would allow for greater flexibility and would adhere to the user's defined naming conventions across all collections. It also opens the possibility for the inclusion of other default fields, as defined in the project settings, which improves the adaptability of the Directus platform.

### Must Have:

- Ability to set project-wide default field names under project settings or in the config file. This includes all currently predefined fields.

### Should Have:
- Compatibility with existing default field configurations. This feature should be implemented in a way that has zero impact on users migrating to a new version of Directus.

### Could Have:
- Option to add other "default" fields to the "Create New Collection" form. These fields could be selected on a case-by-case basis when creating a new collection, providing additional customization options for users.

### Won't Have:
- Changes or additions beyond those outlined above. The focus of this feature is solely on enhancing the flexibility and efficiency of field naming during the creation of new collections.

### Drawbacks

If someone can please point out any drawback I would hear that cladly

### Alternatives

The only alternative that comes to mind would be to keep the current implementation of hardcoded values, which has its own drawbacks in terms of lack of flexibility and user customization.

### Adoption Strategy

The adoption of this feature aims to be seamless for existing Directus developers. As proposed, the implementation would default to the existing hardcoded values when no project-wide settings are configured. This approach ensures zero impact on current Directus projects and users during the transition period.

Because the proposed feature enhances functionality rather than altering existing structures, it should not be considered a breaking change. A migration wouldn't be necessary, as the new functionality would simply offer an additional layer of customization for those who choose to use it.

As for its effect on the broader Directus ecosystem, the proposed feature offers potential benefits without disrupting existing practices. It aligns with the overall Directus philosophy of flexibility and customization, potentially contributing to improved user satisfaction.

This feature could be introduced as a patch following Semantic Versioning (Semver) principles, as it essentially completes an aspect of the system that wasn't fully implemented in the initial design. ;)

**Comment by connorwinston** [comment]
Hi! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by mkungla** [reply]
Has there been a meeting to review this request?

**Reply by rijkvanzanten** [reply]
Not yet! We're trying to go "top to bottom" sorted by upvotes, so we'll get to this one shortly 🙂

**Reply by mkungla** [reply]
still pending?

**Comment by mkungla** [comment]
Over year has passed - there seems to be a little problem how proposals are tracked!

**Reply by rijkvanzanten** [reply]
How so? This is an open feature request still 🙂 We're a relatively small team building this thing and have to prioritize our time accordingly

**Comment by Stretsh** [comment]
How high of a priority does this have? In other words, is it naive to expect it soon? :)

One addition: I would be happy enough (maybe even prefer) to have a config file (JSON, CSV, or JS object, doesn't matter) that I can fill in manually, instead of using the UI. Even if I then have to import it into the system from the UI (or it's done during install) I wouldn't mind.

But I think the core for any of this to work is what this FR proposes.
